File: 003/main.c
#include <assert.h>
#include <math.h>
#include <stdio.h>

#define ll long long

ll smallest_factor(ll n) {
  assert(n >= 2);
  ll i;
  for (i = 2; i < sqrt(n) + 1; ++i)
    if (n % i == 0)
      return i;
  return n;
}

int main(void) {
  ll n = 600851475143;

  for (;;) {
    ll f = smallest_factor(n);
    if (n > f)
      n = n / f;
    else
      break;
  }

  printf("%lld\n", n);

  return 0;
}

File: 003/main.py
import math


def smallest_factor(n: int) -> int:
    assert n >= 2
    for i in range(2, math.isqrt(n) + 1):
        if n % i == 0:
            return i
    return n


while True:
    n = 600851475143
    f = smallest_factor(n)
    if n > f:
        n //= f
    else:
        print(n)
        break
